src/types/bytearray.rs

```diff
@@ -200,7 +202,9 @@ impl PyByteArray {
     /// apply to this function as well.
     #[allow(clippy::mut_from_ref)]
     pub unsafe fn as_bytes_mut(&self) -> &mut [u8] {
-        slice::from_raw_parts_mut(self.data(), self.len())
+        let slice = Py2::borrowed_from_gil_ref(&self).as_bytes_mut();
+        // SAFETY: &self guarantees the reference is alive long enough
+        unsafe { slice::from_raw_parts_mut(slice.as_mut_ptr(), slice.len()) }
```

Comment:
This is unfortunate. Can we maybe generalize the reasoning? For example into something like

```rust
pub fn borrowed_from_gil_ref(gil_ref: &'py T::AsRefTarget) -> &'py Self;
```

to at least produce `&'py Py2<'py, T::AsRefTarget>` to avoid having to do this in each implementation? Is that even valid? I guess the GIL pool should keep the reference alive for the whole GIL lifetime, shouldn't it?

Reply:
Not quite like that because `&'py gil_ref` is a pointer, but `&'py Self` is a pointer to a pointer. I think we might be able to do that with a closure?

```rust
pub fn borrow_from_gil_ref<R>(gil_ref: &'py T::AsRefTarget, f: impl FnOnce(&'py Self) -> R) -> R;
```

I also have a type `Py2Borrowed<'a, 'py, T>(NonNull<ffi::PyObject>, PhantomData<&'a Py2<'py, T>>)` which _is_ pointer layout and necessary to avoid introducing reference count increases on e.g. `PyTuple::get_item`. That can implement `from_gil_ref(gil_ref: &'py T::AsRefTarget) -> Py2Borrowed<'py, 'py, T>;` soundly. Let me also push that into a separate PR so you can browse it.

Reply:
Ah no, the closure doesn't work, because of `'py` lifetime on the reference it can legally escape the closure:

```rust
let mut borrow = None;
Py2::<PyBytes>::borrow_from_gil_ref(&self, |b| {
    borrow = Some(b);
});
dbg!(borrow);  // this blows up
```

Reply:
Ah, I think I see a way to make this work using `Py2Borrowed` as a base implementation, I'll try to push that later this evening after my son's bedtime...